File: src/fft.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h> 


typedef struct {
    double real;
    double imag;
} Complex;


Complex complex_add(Complex a, Complex b) {
    Complex result = {a.real + b.real, a.imag + b.imag};
    return result;
}

Complex complex_sub(Complex a, Complex b) {
    Complex result = {a.real - b.real, a.imag - b.imag};
    return result;
}

Complex complex_mul(Complex a, Complex b) {
    Complex result = {
        a.real * b.real - a.imag * b.imag,
        a.real * b.imag + a.imag * b.real
    };
    return result;
}

Complex complex_exp(double angle) {
    Complex result = {cos(angle), sin(angle)};
    return result;
}

void dft(Complex* x, Complex* X, int N) {
    for (int k = 0; k < N; k++) {
        double sum_real = 0.0;
        double sum_imag = 0.0;
        for (int n = 0; n < N; n++) {
            double angle = -2.0 * M_PI * k * n / N;
            sum_real += x[n].real * cos(angle) - x[n].imag * sin(angle);
            sum_imag += x[n].real * sin(angle) + x[n].imag * cos(angle);
        }
        X[k].real = sum_real;
        X[k].imag = sum_imag;
    }
}

void fft_radix2(Complex* x, Complex* X, int N) {
    if (N == 1) {
        X[0] = x[0];
        return;
    }

    Complex* even = (Complex*)malloc(N / 2 * sizeof(Complex));
    Complex* odd  = (Complex*)malloc(N / 2 * sizeof(Complex));
    Complex* even_fft = (Complex*)malloc(N / 2 * sizeof(Complex));
    Complex* odd_fft  = (Complex*)malloc(N / 2 * sizeof(Complex));

    for (int i = 0; i < N / 2; i++) {
        even[i] = x[2 * i];
        odd[i]  = x[2 * i + 1];
    }

    fft_radix2(even, even_fft, N / 2);
    fft_radix2(odd,  odd_fft,  N / 2);

    for (int k = 0; k < N / 2; k++) {
        double angle = -2.0 * M_PI * k / N;
        Complex twiddle = complex_exp(angle);
        Complex t = complex_mul(twiddle, odd_fft[k]);
        X[k] = complex_add(even_fft[k], t);
        X[k + N / 2] = complex_sub(even_fft[k], t);
    }

    free(even);
    free(odd);
    free(even_fft);
    free(odd_fft);
}

unsigned int bit_reverse(unsigned int x, int log2n) {
    unsigned int n = 0;
    for (int i = 0; i < log2n; i++) {
        n <<= 1;
        n |= (x & 1);
        x >>= 1;
    }
    return n;
}

void fft_radix2_iterative(Complex* x, int N) {
    int log2n = 0;
    int n = N;
    while (n >>= 1) log2n++;

    // Bit reversal
    for (int i = 0; i < N; i++) {
        int j = bit_reverse(i, log2n);
        if (j > i) {
            Complex temp = x[i];
            x[i] = x[j];
            x[j] = temp;
        }
    }

    // FFT
    for (int s = 1; s <= log2n; s++) {
        int m = 1 << s;
        double angle_step = -2.0 * M_PI / m;
        for (int k = 0; k < N; k += m) {
            for (int j = 0; j < m / 2; j++) {
                double angle = j * angle_step;
                Complex twiddle = complex_exp(angle);
                Complex t = complex_mul(twiddle, x[k + j + m / 2]);
                Complex u = x[k + j];
                x[k + j] = complex_add(u, t);
                x[k + j + m / 2] = complex_sub(u, t);
            }
        }
    }
}

void rfft(double* x_real, Complex* X_rfft, int N) {
    Complex* x_complex = (Complex*)malloc(N * sizeof(Complex));
    Complex* X_full = (Complex*)malloc(N * sizeof(Complex));

    for (int i = 0; i < N; i++) {
        x_complex[i].real = x_real[i];
        x_complex[i].imag = 0.0;
    }

    dft(x_complex, X_full, N);

    // Copia só os N/2 + 1 primeiros coeficientes
    for (int i = 0; i <= N / 2; i++) {
        X_rfft[i] = X_full[i];
    }

    free(x_complex);
    free(X_full);
}

double fft_bin_to_hz(int k, double fs, int N) {
    return (fs * k) / N;
}

File: utils/fft.py
import math
import cmath


def fft_radix2(x):
    """
    Implementa a FFT radix-2 Cooley-Tukey recursiva.
    Entrada:
        x: Lista de números complexos (tamanho N, potência de 2)
    Saída:
        Lista com a FFT de x (tamanho N)
    """
    N = len(x)
    if N <= 1:
        return x

    if N % 2 != 0:
        raise ValueError('O tamanho da entrada deve ser uma potência de 2')

    # Divide em pares e ímpares
    even = fft_radix2(x[0::2])
    odd = fft_radix2(x[1::2])

    # Combina as partes
    result = [0] * N
    for k in range(N // 2):
        twiddle = cmath.exp(-2j * cmath.pi * k / N) * odd[k]
        result[k] = even[k] + twiddle
        result[k + N // 2] = even[k] - twiddle

    return result


def fft_radix2_iterative(x):
    """
    Implementa a FFT radix-2 Cooley-Tukey iterativa.
    Entrada:
        x: Lista de números complexos (tamanho N, potência de 2)
    Saída:
        Lista com a FFT de x (tamanho N)
    """
    N = len(x)
    if N <= 1:
        return x

    if N % 2 != 0:
        raise ValueError('O tamanho da entrada deve ser uma potência de 2')

    # Reordena os dados (bit-reversal)
    indices = [0] * N
    for i in range(N):
        indices[i] = int(
            '{:0{width}b}'.format(i, width=N.bit_length() - 1)[::-1], 2
        )

    x_reordered = [x[i] for i in indices]

    # FFT iterativa
    step = 1
    while step < N:
        for k in range(0, N, step * 2):
            for j in range(step):
                twiddle = cmath.exp(-2j * cmath.pi * j / (step * 2))
                t = twiddle * x_reordered[k + j + step]
                x_reordered[k + j + step] = x_reordered[k + j] - t
                x_reordered[k + j] += t
        step *= 2

    return x_reordered


def dft(x):
    """
    Calcula a DFT de um sinal x usando a definição direta.
    Entrada:
        x: Lista de números reais (ou complexos), tamanho N
    Saída:
        X: Lista de números complexos, tamanho N (DFT completa)
    """
    N = len(x)
    X = []
    for k in range(N):
        sum_real = 0.0
        sum_imag = 0.0
        for n in range(N):
            angle = -2 * math.pi * k * n / N
            sum_real += x[n] * math.cos(angle)
            sum_imag += x[n] * math.sin(angle)
        X.append(complex(sum_real, sum_imag))
    return X


def rfft(x):
    """
    Real FFT: calcula apenas os coeficientes únicos da DFT de um sinal real.
    Entrada:
        x: Lista de números reais
    Saída:
        X_rfft: Lista de números complexos contendo os N//2 + 1 primeiros coeficientes (DC até Nyquist)
    """
    N = len(x)
    X_full = dft(x)          # Calcula a DFT completa
    print(f'DFT completa (tamanho {N}): {X_full}')
    X_rfft = X_full[
        : N // 2 + 1
    ]   # Mantém só a metade positiva (inclui DC e Nyquist se N par)
    print(f'RFFT (tamanho {len(X_rfft)}): {X_rfft}')
    return X_rfft


# Exemplo de teste
if __name__ == '__main__':
    # Sinal de exemplo: senoide simples de 4 pontos
    import matplotlib.pyplot as plt

    fs = 8  # Hz
    t = [n / fs for n in range(8)]  # 8 amostras
    x = [math.sin(2 * math.pi * 1 * tn) for tn in t]  # Senoide de 1 Hz

    X_rfft = rfft(x)

    # Exibir resultados
    print('Entrada x:')
    print(x)

    print('\nSaída RFFT (magnitude):')
    for k, Xk in enumerate(X_rfft):
        print(f'k={k}, |X|={abs(Xk):.4f}, Fase={cmath.phase(Xk):.4f} rad')

    # Plot (opcional)
    mags = [abs(Xk) for Xk in X_rfft]
    plt.stem(range(len(mags)), mags)
    plt.title('Magnitude da RFFT')
    plt.xlabel('k (bin)')
    plt.ylabel('|X(k)|')
    plt.show()
